src/main/java/net/snowflake/ingest/streaming/internal/ExternalVolume.java

```diff
@@ -1,9 +1,409 @@
+                  EntityUtils.toString(response.getEntity())));
+
+      client.handleStorageException(ex, 0, "upload", null, null, null);
+    }
+  }
+
+  private void addHeadersToHttpRequest(
+      HttpPut httpRequest, byte[] blob, StageInfo stageInfo, SnowflakeStorageClient client) {
+    // no need to set this as it causes a Content-length header is already set error in apache's
+    // http client.
+    // httpRequest.setHeader("Content-Length", "" + blob.length);
+
+    // TODO: These custom headers need to be a part of the presigned URL HMAC computation in S3,
+    // we'll disable them for now until we can do presigned URL generation AFTER we have the digest.
+
+    /*
+    final String clientKey = this.clientPrefix;
+    final String clientName = this.clientName;
+
+    final byte[] digestBytes;
+    try {
+        digestBytes = MessageDigest.getInstance("SHA-256").digest(blob);
+    } catch (NoSuchAlgorithmException e) {
+        throw new SFException(e, ErrorCode.INTERNAL_ERROR);
+    }
+
+    final String digest = Base64.getEncoder().encodeToString(digestBytes);
+
+    StorageObjectMetadata meta = StorageClientFactory.getFactory().createStorageMetadataObj(stageInfo.getStageType());
+
+    client.addDigestMetadata(meta, digest);
+    client.addStreamingIngestMetadata(meta, clientName, clientKey);
+
+    switch (stageInfo.getStageType()) {
+        case S3:
+            httpRequest.setHeader("x-amz-server-side-encryption", ObjectMetadata.AES_256_SERVER_SIDE_ENCRYPTION);
+            httpRequest.setHeader("x-amz-checksum-sha256", digest); // TODO why does JDBC use a custom x–amz-meta-sfc-digest header for this
+            for (Map.Entry<String, String> entry : meta.getUserMetadata().entrySet()) {
+                httpRequest.setHeader("x-amz-meta-" + entry.getKey(), entry.getValue());
+            }
+
+            break;
+
+        case AZURE:
+            for (Map.Entry<String, String> entry : meta.getUserMetadata().entrySet()) {
+                httpRequest.setHeader("x-ms-meta-" + entry.getKey(), entry.getValue());
+            }
+            break;
+
+        case GCS:
+            for (Map.Entry<String, String> entry : meta.getUserMetadata().entrySet()) {
+                httpRequest.setHeader("x-goog-meta-" + entry.getKey(), entry.getValue());
+            }
+            break;
+    }
+    */
+  }
+
+  PresignedUrlInfo dequeueUrlInfo() {
+    PresignedUrlInfo info = this.presignedUrlInfos.poll();
+    boolean generate = false;
+    if (info == null) {
+      generate = true;
+    } else {
+      // Since the queue had a non-null entry, there is no way numUrlsInQueue is <=0
+      int remainingUrlsInQueue = this.numUrlsInQueue.decrementAndGet();
+      if (remainingUrlsInQueue <= LOW_WATERMARK_FOR_EARLY_REFRESH) {
+        generate = true;
+        // assert remaininUrlsInQueue >= 0
+      }
+    }
+    if (generate) {
+      // TODO: do this generation on a background thread to allow the current thread to make
```

Comment:
I feel like the background thread approach would also be simpler - you won't need all the Semaphore business. The bg thread could just keep refreshing in a loop whenever it sees the remaining urls hit a low.

Reply:
Yes, couple reasons I kept this as simple as possible:

1. There can be many many external volumes (one per table), and the right fix there would be to have a small threadpool that multiplexes over all the active external volumes, and potentially even does a batched retrieval instead of one RPC per ExtVol. This is doable, just a matter of work ordering

2. rate matching - some ext vols will need 1 URL every minute, other volumes will need 10 URLs every second; to do a proper fix the low watermark needs to be adaptive instead of the hardcoded 5 right now.

3. Multipart upload support - to do multipart upload, each part requires its' own presigned URL AND needs to be associated with an upload-id; however now we can't cache URLs in advance as we won't know (in advance) how many parts any given file will have and don't have a batch of upload-ids either to quickly choose from.

Of these three, ended up just having a roundabout fix for (2) by allowing concurrent URL retrieval, as soon as there's high pressure on URL retrieval we'll end up caching a lot of URLs and reduce pressure.